// @glimpse/glimpse-node-agent/release/inspectors/ExpressInspectorActionRouteView.js
'use strict';
var GuidHelper_1 = require('./../util/GuidHelper');
var ExpressProxy_1 = require('./../tracing/proxies/ExpressProxy');
var HttpHelper_1 = require('./util/HttpHelper');
var Tracing_1 = require('./../tracing/Tracing');
var path = require('path');
var ExpressInspectorActionRouteView = (function () {
    function ExpressInspectorActionRouteView() {
    }
    ExpressInspectorActionRouteView.prototype.init = function (broker, contextManager, dateTime, errorReportingService, scriptManager) {
        var _this = this;
        this.broker = broker;
        this.contextManager = contextManager;
        this.dateTime = dateTime;
        this.errorReportingService = errorReportingService;
        this.scriptManager = scriptManager;
        Tracing_1.default.onAlways(ExpressProxy_1.ExpressProxy.EVENT_INVOKE_PRE_EXPRESS_ROUTE_DISPATCH, function (event) {
            _this.onRouteDispatch(event.data.originalThis);
        });
        Tracing_1.default.onAlways(ExpressProxy_1.ExpressProxy.EVENT_INVOKE_PRE_EXPRESS_RESPONSE_RENDER, function (event) {
            // TODO: https://github.com/Glimpse/Glimpse.Node.Prototype/issues/320 - test missing
            _this.onResponseRender(event.data.originalThis);
        });
        Tracing_1.default.onAlways(ExpressProxy_1.ExpressProxy.EVENT_INVOKE_PRE_EXPRESS_RESPONSE_SEND, function (event) {
            // TODO: https://github.com/Glimpse/Glimpse.Node.Prototype/issues/320 - test missing
            _this.onResponseSend(event.data.originalThis, event.data.originalArgs);
        });
        Tracing_1.default.onAlways(ExpressProxy_1.ExpressProxy.EVENT_INVOKE_PRE_EXPRESS_RESPONSE_END, function (event) {
            // TODO: https://github.com/Glimpse/Glimpse.Node.Prototype/issues/320 - test missing
            _this.onResponseEnd(event.data.originalThis);
        });
        Tracing_1.default.onAlways(ExpressProxy_1.ExpressProxy.EVENT_NOTIFY_EXPRESS_RENDER_COMPLETE, function (event) {
            if (event.data.err) {
                _this.onRenderCompleteError(event.data.err);
            }
            else {
                // TODO: https://github.com/Glimpse/Glimpse.Node.Prototype/issues/320 - test missing
                _this.onRenderComplete();
            }
        });
        Tracing_1.default.onAlways(ExpressProxy_1.ExpressProxy.EVENT_INVOKE_PRE_EXPRESS_VIEW_RENDER, function (event) {
            _this.onViewRender(event.data.originalThis);
        });
    };
    ExpressInspectorActionRouteView.prototype.onRouteDispatch = function (route) {
        // TODO: figure out what correct data to display here for a "route" and an "action"
        var middlewareFunction = route.stack[0].handle;
        var functionName;
        if (middlewareFunction.glimpse) {
            functionName = middlewareFunction.glimpse.originalName || '<anonymous>';
        }
        else {
            functionName = middlewareFunction.name || '<anonymous>';
        }
        var actionId = GuidHelper_1.GuidHelper.newGuid();
        var actionName = functionName;
        var actionDisplayName = functionName;
        var actionControllerName = functionName;
        var actionStartTime = new Date();
        this.broker.sendMessage({
            actionId: actionId,
            actionName: actionName,
            actionDisplayName: actionDisplayName,
            actionControllerName: actionControllerName,
            // 'begin-action', 'action-route'
            actionStartTime: actionStartTime,
            routeName: functionName,
            routePattern: route.path,
            routeData: {
                controller: actionControllerName,
                action: actionName
            },
            // 'before-action-invoked', 'action-content'
            actionTargetClass: '',
            actionTargetMethod: '',
            actionInvokedStartTime: actionStartTime,
            /* tslint:disable no-null-keyword */
            binding: null
        }, ['begin-action', 'action-route', 'before-action-invoked', 'action-content']);
        // Save some action-related context for response rendering...
        var extendedContext = this.contextManager.currentContext();
        extendedContext.action = {
            actionId: actionId,
            actionName: actionName,
            actionControllerName: actionControllerName,
            actionStartTime: actionStartTime,
            beforeActionInvokedSent: true,
            afterActionInvokedSent: false
        };
    };
    ExpressInspectorActionRouteView.prototype.onResponseEnd = function (response) {
        var context = HttpHelper_1.HttpHelper.getContext(response);
        this.contextManager.checkContextID('ExpressInspectorActionRouteView::onResponseEnd()', context ? context.id : undefined);
        if (!context) {
            return;
        }
        var actionContext = context.action;
        // ensure that we send after-action-invoked if it hasn't already been sent.
        // TODO:  There's probably a better way to do this.
        if (actionContext && actionContext.beforeActionInvokedSent && !actionContext.afterActionInvokedSent) {
            var startTime = new Date();
            this.broker.sendMessage({
                actionId: actionContext.actionId,
                actionName: actionContext.actionName,
                actionControllerName: actionContext.actionControllerName,
                // 'after-action-invoked'
                actionInvokedEndTime: startTime,
                actionInvokedDuration: startTime.getTime() - actionContext.actionStartTime.getTime(),
                actionInvokedOffset: actionContext.actionStartTime.getTime() - context.startTime.getTime()
            }, ['after-action-invoked'], undefined, context);
            actionContext.afterActionInvokedSent = true;
        }
    };
    ExpressInspectorActionRouteView.prototype.onRenderCompleteError = function (err) {
        var context = this.contextManager.currentContext();
        this.contextManager.checkContextID('ExpressInspectorActionRouteView::onRenderCompleteError()');
        if (!context) {
            return;
        }
        var actionContext = context.action;
        if (!actionContext) {
            return;
        }
        if (err.view) {
            var viewPath = err.view.path;
            if (!viewPath) {
                viewPath = path.join(err.view.root, err.view.name) + err.view.ext;
            }
            var viewFound = err.view.path ? true : false;
            ExpressInspectorActionRouteView.sendActionViewFound(this.broker, this.dateTime, err.view.name, viewPath, viewFound);
        }
    };
    ExpressInspectorActionRouteView.prototype.onRenderComplete = function () {
        this.contextManager.checkContextID('ExpressInspectorActionRouteView::onRenderComplete()');
        var context = this.contextManager.currentContext();
        if (!context) {
            return;
        }
        var actionContext = context.action;
        if (!actionContext) {
            return;
        }
        var startTime = actionContext.responseRenderStartTime;
        var endTime = new Date();
        this.broker.sendMessage({
            actionId: actionContext.actionId,
            actionName: actionContext.actionName,
            actionControllerName: actionContext.actionControllerName,
            // 'after-action-view-invoked'
            viewEndTime: endTime,
            viewDuration: endTime.getTime() - startTime.getTime(),
            viewOffset: startTime.getTime() - context.startTime.getTime(),
            // 'after-action'
            actionEndTime: endTime,
            actionDuration: endTime.getTime() - actionContext.actionStartTime.getTime(),
            actionOffset: actionContext.actionStartTime.getTime() - context.startTime.getTime()
        }, ['after-action-view-invoked', 'after-action'], undefined, context);
    };
    ExpressInspectorActionRouteView.prototype.onResponseRender = function (response) {
        var context = HttpHelper_1.HttpHelper.getContext(response);
        this.contextManager.checkContextID('ExpressInspectorActionRouteView::onResponseRender()', context ? context.id : undefined);
        if (!context) {
            return;
        }
        var actionContext = context.action;
        if (!actionContext) {
            return;
        }
        var startTime = new Date();
        actionContext.responseRenderStartTime = startTime;
        this.broker.sendMessage({
            actionId: actionContext.actionId,
            actionName: actionContext.actionName,
            actionControllerName: actionContext.actionControllerName,
            // 'after-action-invoked'
            actionInvokedEndTime: startTime,
            actionInvokedDuration: startTime.getTime() - actionContext.actionStartTime.getTime(),
            actionInvokedOffset: actionContext.actionStartTime.getTime() - context.startTime.getTime(),
            // 'before-action-view-invoked'
            viewStartTime: startTime
        }, ['after-action-invoked', 'before-action-view-invoked'], undefined, context);
        actionContext.afterActionInvokedSent = true;
    };
    ExpressInspectorActionRouteView.prototype.onViewRender = function (view) {
        var viewPath = view.path || path.join(view.root, view.name) + view.ext;
        var foundView = view.path ? true : false;
        ExpressInspectorActionRouteView.sendActionViewFound(this.broker, this.dateTime, view.name, viewPath, foundView);
    };
    ExpressInspectorActionRouteView.prototype.onResponseSend = function (response, originalArgs) {
        // OK to continue if no context, we'll just have an empty guid for the context ID in the HUD
        var context = HttpHelper_1.HttpHelper.getContext(response);
        this.contextManager.checkContextID('ExpressInspectorActionRouteView::onResponseSend()', context ? context.id : undefined);
        var position = 0;
        if (originalArgs.length === 2 && typeof originalArgs[0] === 'number' && typeof originalArgs[1] !== 'number') {
            position = 1;
        }
        else if (originalArgs.length === 1 && typeof originalArgs[0] === 'number') {
            position = -1;
        }
        // TODO: this should be an extension point
        if (position >= 0 && typeof originalArgs[position] === 'string' && response.req.baseUrl !== '/glimpse') {
            var contentType = response.get('Content-Type'); // NOTE: Get is case-insensitive.
            if (contentType === undefined || contentType === 'text/html') {
                var scripts = this.scriptManager.getScriptTagsForCurrentRequest(context);
                var body = originalArgs[position];
                body = this.scriptManager.injectScript(body, scripts);
                originalArgs[position] = body;
            }
        }
    };
    ExpressInspectorActionRouteView.sendActionViewFound = function (broker, dateTime, viewName, viewPath, viewFound) {
        var viewSearchedTime = dateTime.now;
        broker.sendMessage({
            // 'action-view-found'
            viewName: viewName,
            viewPath: viewPath,
            viewDidFind: viewFound,
            viewSearchedTime: viewSearchedTime
        }, ['action-view-found']);
    };
    return ExpressInspectorActionRouteView;
}());
exports.ExpressInspectorActionRouteView = ExpressInspectorActionRouteView;
