// @glimpse/glimpse-node-agent/release/inspectors/ExpressInspectorMiddleware.js
'use strict';
var MiddlewareWrapper_1 = require('./util/MiddlewareWrapper');
var Tracing_1 = require('./../tracing/Tracing');
var ExpressProxy_1 = require('./../tracing/proxies/ExpressProxy');
var ExpressInspectorMiddleware = (function () {
    function ExpressInspectorMiddleware() {
    }
    ExpressInspectorMiddleware.prototype.init = function (broker, contextManager, dateTime) {
        var _this = this;
        this.broker = broker;
        this.contextManager = contextManager;
        this.dateTime = dateTime;
        Tracing_1.default.onAlways(ExpressProxy_1.ExpressProxy.EVENT_INVOKE_PRE_EXPRESS_ROUTE_METHOD, function (event) {
            _this.onMiddlewareMethodInvoked(event.data.originalThis.path, event.data.originalArgs, event.data.methodName);
        });
        Tracing_1.default.onAlways(ExpressProxy_1.ExpressProxy.EVENT_INVOKE_PRE_EXPRESS_ROUTER_USE, function (event) {
            _this.onUseInvokedBegin(event.data.originalThis, event.data.originalArgs);
        });
        Tracing_1.default.onAlways(ExpressProxy_1.ExpressProxy.EVENT_INVOKE_PRE_EXPRESS_ROUTER, function (event) {
            _this.onRouterInvokedEnd(event.data.router);
        });
        Tracing_1.default.onAlways(ExpressProxy_1.ExpressProxy.EVENT_INVOKE_PRE_EXPRESS_STATIC, function (event) {
            _this.onStaticInvokedEnd(event.data.middleware);
        });
    };
    ExpressInspectorMiddleware.prototype.onMiddlewareMethodInvoked = function (path, originalArgs, method) {
        if (originalArgs && originalArgs.length === 1 && typeof originalArgs[0] === 'function') {
            originalArgs[0] = MiddlewareWrapper_1.MiddlewareWrapper.wrap(this.broker, this.contextManager, this.dateTime, [path], method !== 'all' ? method.toUpperCase() : undefined, originalArgs[0]);
        }
    };
    ExpressInspectorMiddleware.prototype.onUseInvokedBegin = function (originalThis, originalArgs) {
        if (originalThis.glimpse && originalThis.glimpse.ignore) {
            return;
        }
        if (originalArgs && originalArgs.length >= 1) {
            var paths = ['/'];
            if (originalArgs.length > 1) {
                var firstArg = originalArgs[0];
                if (Array.isArray(firstArg) && firstArg.length > 0 && typeof firstArg[0] === 'string') {
                    paths = firstArg;
                }
                else if (typeof firstArg === 'string') {
                    paths = [firstArg];
                }
                else if (firstArg instanceof RegExp) {
                    paths = [firstArg.toString()];
                }
            }
            else if (originalArgs.length === 1) {
                var middleware = originalArgs[0];
                if (typeof middleware === 'function') {
                    if (!originalThis.glimpse) {
                        originalThis.glimpse = {
                            firstUse: middleware
                        };
                    }
                    else if (!originalThis.glimpse.firstUse) {
                        originalThis.glimpse.firstUse = middleware;
                    }
                    else if (!originalThis.glimpse.secondUse) {
                        originalThis.glimpse.secondUse = middleware;
                        // NOTE: If the first and second middleware added to a Router are 'query' and 'expressInit',
                        //       respectively, it's relatively certain that it's the built-in Express middleware that
                        //       we can't otherwise patch in order to attach metadata, so we do so here.
                        if (originalThis.glimpse.firstUse.name === 'query' && originalThis.glimpse.secondUse.name === 'expressInit') {
                            MiddlewareWrapper_1.MiddlewareWrapper.attachMetadata(originalThis.glimpse.firstUse, 'query', 'Express Query Parser', 'express');
                            MiddlewareWrapper_1.MiddlewareWrapper.attachMetadata(originalThis.glimpse.secondUse, 'expressInit', 'Express Initialization', 'express');
                        }
                    }
                }
            }
            this.wrapNestedMiddleware(originalArgs, paths);
        }
    };
    ExpressInspectorMiddleware.prototype.wrapNestedMiddleware = function (array, paths) {
        for (var i = 0; i < array.length; i++) {
            var arg = array[i];
            if (typeof arg === 'function') {
                if (arg.glimpse && arg.glimpse.ignore) {
                    continue;
                }
                array[i] = MiddlewareWrapper_1.MiddlewareWrapper.wrap(this.broker, this.contextManager, this.dateTime, paths, /* method */ undefined, arg);
            }
            else if (Array.isArray(arg)) {
                this.wrapNestedMiddleware(arg, paths);
            }
        }
    };
    ExpressInspectorMiddleware.prototype.onRouterInvokedEnd = function (router) {
        MiddlewareWrapper_1.MiddlewareWrapper.attachMetadata(router, 'router', 'Express Router', 'express');
    };
    ExpressInspectorMiddleware.prototype.onStaticInvokedEnd = function (middleware) {
        MiddlewareWrapper_1.MiddlewareWrapper.attachMetadata(middleware, 'serveStatic', 'Express Static File Server', 'express');
    };
    return ExpressInspectorMiddleware;
}());
exports.ExpressInspectorMiddleware = ExpressInspectorMiddleware;
