// @glimpse/glimpse-node-agent/inspectors/EnvironmentInspector.ts
'use strict';

import { IncomingMessage, ServerResponse } from 'http';

import { IAgent } from '../IAgent';
import { IServerRequestInspector } from './IServerRequestInspector';
import { HttpHelper } from './util/HttpHelper';

import os = require('os');

export class EnvironmentInspector implements IServerRequestInspector {
    private agent: IAgent;

    public requestStart(req: IncomingMessage, res: ServerResponse) {
        // no-op
    }

    public requestEnd(req: IncomingMessage, res: ServerResponse) {

        const context = HttpHelper.getContext(req);
        if (context) {

            // TODO: Should we be using a start time that's part of the context?
            const now = this.agent.providers.dateTime.now;
            const currentMoment = now.getMoment();

            this.agent.broker.sendMessage(
                {
                    serverName: os.hostname(),
                    serverTime: now.format(false),
                    serverTimezoneOffset: now.getUTCOffset(true),
                    serverDaylightSavingTime: currentMoment.isDST(),
                    serverOperatingSystem: os.type(),
                    runtimeVersion: process.version,
                    runtimeArchitecture: os.arch()
                },
                ['environment'],
                undefined,
                context
            );
        }
    }

    public responseEnd(req: IncomingMessage, res: ServerResponse) {
        // TODO: Remove once http-proxy allows for removal.
    }

    public init(agent: IAgent) {
        this.agent = agent;
    }
}
