// @glimpse/glimpse-node-agent/inspectors/KnexProxy.ts
'use strict';

import { IAgent } from '../IAgent';
import { ProxyBase } from './ProxyBase';

export class KnexProxy extends ProxyBase {
    private setupProxy(agent: IAgent, knex) {
        return function knexProxy() {
            const realKnex = knex.apply(this, arguments);

            const onStart = function onKnexStart(builder) {

                const context = agent.providers.contextManager.currentContext();

                // NOTE: If there is no current context, the queries are being made outside of any outstanding
                //       HTTP request (such as during application startup) and these should not be logged.

                if (context) {
                    const startTime = new Date();
                    const queries = [];

                    builder.on('query', function onKnexQuery(query) {
                        queries.push(query);

                        agent.broker.sendMessage(
                            {
                                commandText: query.sql,
                                commandType: '',
                                commandParameters: '',
                                commandMethod: '',
                                commandIsAsync: true,
                                commandStartTime: startTime
                            },
                            ['before-execute-command']);
                    });

                    builder.on('end', function onKnexEnd() {
                        const endTime = new Date();

                        queries.forEach(function forQuery() {

                            agent.broker.sendMessage(
                                {
                                    commandHadException: false,
                                    commandEndTime: endTime,
                                    commandDuration: endTime.getTime() - startTime.getTime(),
                                    commandOffset: startTime.getTime() - context.startTime.getTime()
                                },
                                ['after-execute-command']);
                        });
                    });
                }
            };

            realKnex.client.on('start', onStart);

            return realKnex;
        };
    };

    public get moduleName() { return 'knex'; };

    public init(agent: IAgent, knex) {
        return this.setupProxy(agent, knex);
    };

    public get forceLoadModule(): boolean {
        return false;
    }
}
