// @glimpse/glimpse-node-agent/configuration/ResourceProvider.ts
'use strict';

import { IContextManager } from '../messaging/IContextManager';
import { IResourceProvider } from './IResourceProvider';
import { ResourceProviderHttp } from './ResourceProviderHttp';
import { ResourceProviderInProc } from './ResourceProviderInProc';

export interface IResourceProviderOptions {
    metadataUri?: string;
    server?;
}

export class ResourceProvider implements IResourceProvider {
    private provider: IResourceProvider;

    public getResourceDefinitions(callback: (resources: { [key: string]: string }) => void): void {
        this.provider.getResourceDefinitions(callback);
    }

    public init(contextManager: IContextManager, options: IResourceProviderOptions): void {
        if (options && options.metadataUri) {
            this.provider = new ResourceProviderHttp(options.metadataUri);
        }
        else if (options && options.server) {
            this.provider = new ResourceProviderInProc(contextManager, options.server);
        }
        else {
            throw new Error('One of the \'metadataUri\' or \'server\' options must be specified.');
        }
    }
}
