// @glimpse/glimpse-node-agent/IAgent.ts
import { IAgentBroker } from './messaging/IAgentBroker';
import { IContextManager } from './messaging/IContextManager';
import { IDateTime } from './configuration/IDateTime';
import { IErrorReportingService } from './errors/IErrorReportingService';
import { IMessagePublisher } from './messaging/IMessagePublisher';
import { IResourceProvider } from './configuration/IResourceProvider';
import { IScriptManager } from './messaging/IScriptManager';
import { IConfigSettings } from './configuration/IConfigSettings';

/**
 * Represents the set of options with which to configure the agent.
 */
export interface IAgentOptions {

    /**
     * (Optional) The URI of the metadata endpoint for the Glimpse server, if not embedded in-process.
     */
    metadataUri?: string;

    /**
     * (Optional) The instance of embedded, in-process Glimpse server.
     */
    server?;
}

/**
 *  Exposes the set of service providers for use by other Glimpse components.
 * 
 *  Glimpse components should use these instances in favor of importing a service provider themselves.
 */
export interface IAgentProviders {
    contextManager: IContextManager;
    dateTime: IDateTime;
    errorReportingService: IErrorReportingService;
    messagePublisher: IMessagePublisher;
    resourceProvider: IResourceProvider;
    scriptManager: IScriptManager;
    configSettings: IConfigSettings;
}

/**
 *  Represents the root Glimpse agent component.
 */
export interface IAgent {
    broker: IAgentBroker;
    providers: IAgentProviders;

    init(options?: IAgentOptions);
}
