// @glimpse/glimpse-node-server/release/resources/MessageStreamResource.js
'use strict';
var Resource = (function () {
    function Resource(server) {
        this.name = 'message-stream';
        this.uriTemplate = '{?types,contextId}';
        this.type = 'client';
        if (server) {
            this.init(server.providers.messagePublisher);
        }
    }
    Resource.prototype.init = function (messagePublisher) {
        this.messagePublisher = messagePublisher;
    };
    Resource.prototype.invoke = function (req, res) {
        // TODO: Still need to complete pings (see #45)
        var types = req.query.types ? req.query.types.split(',') : [];
        var contextId = req.query.contextId;
        res.writeHead(200, {
            'Content-Type': 'text/event-stream',
            'Cache-Control': 'no-cache',
            'Connection': 'keep-alive'
        });
        var subscription = this.messagePublisher.streamMessages(contextId, types, function (err, message) {
            if (err) {
                return err;
            }
            if (message) {
                var payload = '';
                payload += 'id: ' + message.context.id + '\n';
                payload += 'event: ' + 'message' + '\n';
                payload += 'data: ' + '[' + message.payload + ']' + '\n';
                payload += '\n';
                res.write(payload);
            }
            else {
                res.end();
            }
        });
        req.on('close', function done() {
            subscription.done();
        });
    };
    ;
    return Resource;
}());
exports.Resource = Resource;
