// @glimpse/glimpse-node-server/release/resources/MetadataResource.js
'use strict';
var hash_1 = require('../services/hash');
var UriTemplate_1 = require('./UriTemplate');
var _ = require('lodash');
var Resource = (function () {
    function Resource(server) {
        this.name = 'metadata';
        this.uriTemplate = '?hash={hash}';
        this.templateName = 'metadataTemplate';
        this.type = 'client';
        this.resourceManager = server.providers.resourceManager;
    }
    Resource.prototype.invoke = function (req, res) {
        var baseUri = req.protocol + '://' + req.get('host') + '/glimpse';
        var resources = _(this.resourceManager.resources)
            .reduce(function (result, resource) {
            result[resource.name] = UriTemplate_1.UriTemplate.fromResource(baseUri, resource);
            return result;
        }, {});
        var resourceHash = hash_1.Hash.hashObject(resources);
        var metadata = {
            'resources': resources,
            'hash': resourceHash
        };
        res.json(metadata);
    };
    ;
    return Resource;
}());
exports.Resource = Resource;
