// @glimpse/glimpse-node-agent/release/messaging/ContextManagerBase.js
'use strict';
var GuidHelper_1 = require('./../util/GuidHelper');
var ErrorReporting_1 = require('../errors/ErrorReporting');
var Errors = require('../errors/Errors');
var ContextManagerBase = (function () {
    function ContextManagerBase() {
    }
    ContextManagerBase.prototype.init = function () {
        //no-op
    };
    ContextManagerBase.prototype.createContext = function (req) {
        var context = {
            id: GuidHelper_1.GuidHelper.newGuid(false),
            type: 'Request',
            startTime: new Date(),
            req: req
        };
        return context;
    };
    ;
    ContextManagerBase.prototype.isWithinContext = function () {
        return (this.currentContext() ? true : false);
    };
    ContextManagerBase.prototype.currentContext = function () {
        throw new Error('please override');
    };
    ContextManagerBase.prototype.runInNewContext = function (req, callback) {
        throw new Error('please override');
    };
    ContextManagerBase.prototype.wrapInCurrentContext = function (callback) {
        return callback;
    };
    ContextManagerBase.prototype.checkContextID = function (location, expectedContextID) {
        var currentContext = this.currentContext();
        if (expectedContextID) {
            var currentContextID = currentContext ? currentContext.id : undefined;
            if (currentContextID !== expectedContextID) {
                ErrorReporting_1.ErrorReporting.getService().reportError(Errors.unexpectedContext(), location, expectedContextID, currentContextID);
            }
        }
        else if (!currentContext) {
            ErrorReporting_1.ErrorReporting.getService().reportError(Errors.noContext(), location);
        }
    };
    return ContextManagerBase;
}());
exports.ContextManagerBase = ContextManagerBase;
