// @glimpse/glimpse-node-server/release/Server.js
'use strict';
var MessagePublisher_1 = require('./messaging/MessagePublisher');
var InMemoryMessageQuery_1 = require('./storage/InMemoryMessageQuery');
var InMemoryMessageStorage_1 = require('./storage/InMemoryMessageStorage');
var ResourceManager_1 = require('./resources/ResourceManager');
var Server = (function () {
    function Server() {
        this._messageStorage = new InMemoryMessageStorage_1.InMemoryMessageStorage();
        this._messagePublisher = new MessagePublisher_1.MessagePublisher(this._messageStorage);
        this._messageQuery = new InMemoryMessageQuery_1.InMemoryMessageQuery(this._messageStorage);
        this._resourceManager = new ResourceManager_1.ResourceManager();
        this.providers = {
            messagePublisher: this._messagePublisher,
            messageQuery: this._messageQuery,
            resourceManager: this._resourceManager,
            // TODO: Set this properly once the resource provider has been converted to TypeScript.
            resourceProvider: undefined
        };
    }
    return Server;
}());
exports.Server = Server;
exports.instance = new Server();
