// @glimpse/glimpse-node-agent/release/tracing/proxies/ExpressProxyActionRouteView.js
'use strict';
var ExpressProxy_1 = require('./ExpressProxy');
var Tracing_1 = require('./../Tracing');
var ExpressProxyActionRouteView = (function () {
    function ExpressProxyActionRouteView() {
    }
    ExpressProxyActionRouteView.init = function (express) {
        var expressDispatch = express.Route.prototype.dispatch;
        express.Route.prototype.dispatch = function dispatchRoute() {
            if (Tracing_1.default.isEventEnabled(ExpressProxy_1.ExpressProxy.EVENT_INVOKE_PRE_EXPRESS_ROUTE_DISPATCH)) {
                var data = {
                    originalThis: this,
                    originalArgs: arguments
                };
                Tracing_1.default.publish(ExpressProxy_1.ExpressProxy.EVENT_INVOKE_PRE_EXPRESS_ROUTE_DISPATCH, data);
            }
            expressDispatch.apply(this, arguments);
        };
        function viewRender() {
            if (Tracing_1.default.isEventEnabled(ExpressProxy_1.ExpressProxy.EVENT_INVOKE_PRE_EXPRESS_VIEW_RENDER)) {
                var data = {
                    originalThis: this,
                    originalArgs: arguments
                };
                Tracing_1.default.publish(ExpressProxy_1.ExpressProxy.EVENT_INVOKE_PRE_EXPRESS_VIEW_RENDER, data);
            }
            var expressViewRender = this.originalRender;
            expressViewRender.apply(this, arguments);
        }
        ;
        var expressResponseRender = express.response.render;
        express.response.render = function renderResponse() {
            var rest = [];
            for (var _i = 0; _i < arguments.length; _i++) {
                rest[_i - 0] = arguments[_i];
            }
            // Need to pass original arguments
            var updatedArguments = ExpressProxyActionRouteView.fixupResponseRenderArgs(this, rest);
            // hijack the view render here - A "view engine" is pluggable on the app, so we want to support
            // potential of multiple apps in a process with different "view" settings.
            if (!this.req.app.settings.view.prototype.originalRender) {
                this.req.app.settings.view.prototype.originalRender = this.req.app.settings.view.prototype.render;
                this.req.app.settings.view.prototype.render = viewRender;
            }
            if (Tracing_1.default.isEventEnabled(ExpressProxy_1.ExpressProxy.EVENT_INVOKE_PRE_EXPRESS_RESPONSE_RENDER)) {
                var data = {
                    originalThis: this,
                    originalArgs: updatedArguments
                };
                Tracing_1.default.publish(ExpressProxy_1.ExpressProxy.EVENT_INVOKE_PRE_EXPRESS_RESPONSE_RENDER, data);
            }
            expressResponseRender.apply(this, updatedArguments);
        };
        var expressSend = express.response.send;
        express.response.send = function sendResponse() {
            if (Tracing_1.default.isEventEnabled(ExpressProxy_1.ExpressProxy.EVENT_INVOKE_PRE_EXPRESS_RESPONSE_SEND)) {
                var data = {
                    originalThis: this,
                    originalArgs: arguments
                };
                Tracing_1.default.publish(ExpressProxy_1.ExpressProxy.EVENT_INVOKE_PRE_EXPRESS_RESPONSE_SEND, data);
            }
            expressSend.apply(this, arguments);
        };
        var expressEnd = express.response.end;
        express.response.end = function endResponse() {
            if (Tracing_1.default.isEventEnabled(ExpressProxy_1.ExpressProxy.EVENT_INVOKE_PRE_EXPRESS_RESPONSE_END)) {
                var data = {
                    originalThis: this,
                    originalArgs: arguments
                };
                Tracing_1.default.publish(ExpressProxy_1.ExpressProxy.EVENT_INVOKE_PRE_EXPRESS_RESPONSE_END, data);
            }
            expressEnd.apply(this, arguments);
        };
    };
    ExpressProxyActionRouteView.fixupResponseRenderArgs = function (response, originalArguments) {
        var done;
        function onRenderComplete(err, str) {
            if (Tracing_1.default.isEventEnabled(ExpressProxy_1.ExpressProxy.EVENT_NOTIFY_EXPRESS_RENDER_COMPLETE)) {
                var data = {
                    originalThis: this,
                    originalArgs: arguments,
                    err: err
                };
                Tracing_1.default.publish(ExpressProxy_1.ExpressProxy.EVENT_NOTIFY_EXPRESS_RENDER_COMPLETE, data);
            }
            done(err, str);
        }
        ;
        var updatedArguments = Array.prototype.slice.call(originalArguments);
        if (updatedArguments.length > 0 && (typeof updatedArguments[updatedArguments.length - 1] === 'function')) {
            done = updatedArguments[updatedArguments.length - 1];
            updatedArguments[updatedArguments.length - 1] = onRenderComplete;
        }
        else {
            updatedArguments.push(onRenderComplete);
        }
        // default callback - this is the default "done" handler set up by express response object.  It ensures we have correct error handling. 
        done = done || function (err, str) {
            if (err) {
                return response.req.next(err);
            }
            response.send(str);
        };
        return updatedArguments;
    };
    return ExpressProxyActionRouteView;
}());
exports.ExpressProxyActionRouteView = ExpressProxyActionRouteView;
